cli/pack/app_dir.go

```diff
@@ -106,48 +135,142 @@ func copyFromCache(paths CachePaths, destPath string, ctx *context.Ctx) {
+
+			return false, err
+		}
+	}
+
+	return false, nil
+}
 
+func copyPathFromCache(cachedPath string, destPath string, pathFromRoot string) error {
+	baseDestPath := filepath.Base(destPath)
+	cacheIsFile, err := cachePathIsFile(cachedPath, baseDestPath)
 	if err != nil {
+		return fmt.Errorf("Failed to determine if the path is a file: %s", err)
+	}
+
+	if cacheIsFile {
+		cachedPath = filepath.Join(cachedPath, baseDestPath)
+	}
+
+	if err := copy.Copy(cachedPath, destPath); err != nil {
 		return fmt.Errorf("Failed to copy path %s from cache to project directory: %s", destPath, err)
 	}
 
+	log.Infof("Using cached path %s", pathFromRoot)
 	return nil
 }
 
-func getProjectCachePaths(ctx *context.Ctx) (CachePaths, error) {
-	if ctx.Pack.NoCache {
-		return nil, nil
+func parseCacheParamsFile(cacheParamsPath string) (CachePathsParams, error) {
+	if _, err := os.Stat(cacheParamsPath); os.IsNotExist(err) {
+		return nil, fmt.Errorf("File %s with pack cache paths parameters doesn't exists", cacheParamsPath)
+	} else if err != nil {
+		return nil, fmt.Errorf("Failed to process %s file: %s", cacheParamsPath, err)
 	}
 
-	rockspecPath, err := common.FindRockspec(ctx.Project.Path)
+	fileContent, err := common.GetFileContentBytes(cacheParamsPath)
 	if err != nil {
-		return nil, fmt.Errorf("Unable to find rockspec: %s", err)
-	} else if rockspecPath == "" {
-		return nil, fmt.Errorf("Application directory should contain rockspec")
+		return nil, fmt.Errorf("Failed to read %s file: %s", cacheParamsPath, err)
+	}
+
+	var cachePathsParams CachePathsParams
+	if err := yaml.Unmarshal(fileContent, &cachePathsParams); err != nil {
+		return nil, fmt.Errorf("Failed to parse file %s with pack cache parameters: %s", cacheParamsPath, err)
+	}
+
+	return cachePathsParams, nil
+}
+
+func calculateCachePath(ctx *context.Ctx, params *CachePathParams) (string, error) {
+	var keyHash string
+	projectPathHash := common.StringSHA1Hex(ctx.Project.Path)[:cntFirstSymbolsFromHash]
+
+	switch {
+	case params.AlwaysCache == true:
+		keyHash = "always"
+	case params.KeyPath != "":
+		pathFromProjectRoot := filepath.Join(ctx.Project.Path, params.KeyPath)
+		if _, err := os.Stat(pathFromProjectRoot); err == nil {
+			if keyHash, err = common.FileSHA1Hex(pathFromProjectRoot); err != nil {
+				return "", fmt.Errorf("Failed to get hash from file content for path %s: %s", params.Path, err)
+			}
+		} else if os.IsNotExist(err) {
+			return "", fmt.Errorf("Specified key-path file %s for the path %s does not exist", pathFromProjectRoot, params.Path)
+		} else {
+			return "", fmt.Errorf("Failed to get specified file for path %s: %s", params.Path, err)
+		}
+
+		keyHash = keyHash[:cntFirstSymbolsFromHash]
+	default:
+		keyHash = common.StringSHA1Hex(params.Key)[:cntFirstSymbolsFromHash]
+	}
+
+	return filepath.Join(ctx.Cli.CacheDir, projectPathHash, params.Path, keyHash), nil
+}
+
+func validateCacheParams(params *CachePathParams) error {
+	if params.Key != "" && params.KeyPath != "" {
```

Comment:
Look at `common.OnlyOneIsTrue`.

Reply:
It seems to me that using this function (and especially with ``!`` operator), only complicates the readability of the code.

Reply:
Are you sure?
The whole function can be replaced by

```go
If not common.OnlyOneIsTrue([]bool{params.Key != "", params.KeyPath != "", params.AlwaysCache}) {
    return fmt.Errorf(cacheParamsErrorMsg, params.Path)
}
```

Reply:
Agree